jdeeco-core/src/cz/cuni/mff/d3s/deeco/runtimelog/RuntimeLogger.java

```diff
@@ -0,0 +1,537 @@
+		instance.indexWriter = new BufferedWriter(indexOut);
+		instance.backLogWriter = new BufferedWriter(backLogOut);
+
+		instance.timeProvider = currentTimeProvider;
+		instance.scheduler = scheduler;
+
+		instance.currentDataOffset = 0;
+
+		for (SnapshotProvider sp : instance.snapshotProviders.keySet()) {
+			long period = instance.snapshotProviders.get(sp);
+			RuntimeLogTimerTaskListener slttListener = new RuntimeLogTimerTaskListener(
+					sp, period);
+			scheduler.addTask(new CustomStepTask(scheduler, slttListener,
+					period));
+		}
+		instance.snapshotProviders.clear();
+
+		for (Long time : instance.backLogOffsets) {
+			instance.backLogWriter.write(String.format("%d\n", time));
+		}
+		instance.backLogWriter.flush();
+		instance.backLogOffsets.clear();
+
+	}
+
+	/**
+	 * Provides the singleton instance of the {@link RuntimeWriter}.
+	 * @return The singleton instance of the {@link RuntimeWriter}.
+	 */
+	public static RuntimeLogger getInstance() {
+		return instance;
+	}
+
+	/**
+	 * Writes the data from the given <em>record</em> to the {@link RuntimeLogger#dataWriter}.
+	 * The written record has a format of an <a href="http://en.wikipedia.org/wiki/XML">XML</a>
+	 * element. The values from the given <em>record</em> are stored as attributes of the
+	 * written <a href="http://en.wikipedia.org/wiki/XML">XML</a> element.
+	 * <p> The {@link Writer} is not being flushed when this method is called. </p>
+	 * @param record contains the data that will be logged.
+	 * @throws IOException Thrown if there is a problem writing into the {@link Writer}s
+	 * for the logging.
+	 * @throws IllegalStateException Thrown if the {@link RuntimeLogger#init} method hasn't
+	 * been called before this method is being called.
+	 * @throws IllegalArgumentException Thrown if the given argument is null.
+	 */
+	public static void log(RuntimeLogRecord record) throws IOException,
+			IllegalStateException {
+		log(record, false);
+	}
+
+	/**
+	 * Writes the data from the given <em>snapshotData</em> argument to the {@link RuntimeLogger#dataWriter}.
+	 * The written record has a format of an <a href="http://en.wikipedia.org/wiki/XML">XML</a>
+	 * element. The values from the given <em>snapshotData</em> argument are stored as attributes of the
+	 * written <a href="http://en.wikipedia.org/wiki/XML">XML</a> element. Also an index record
+	 * pointing to this snapshot record is written using the {@link RuntimeLogger#indexWriter}.
+	 * <p> The {@link Writer}s are being flushed when this method is called. </p>
+	 * @param snapshotData contains the data that will be logged.
+	 * @throws IOException Thrown if there is a problem writing into the {@link Writer}s
+	 * for the logging.
+	 * @throws IllegalStateException Thrown if the {@link RuntimeLogger#init} method hasn't
+	 * been called before this method is being called.
+	 * @throws IllegalArgumentException Thrown if the given argument is null.
+	 */
+	public static void logSnapshot(RuntimeLogRecord snapshotData)
```

Comment:
Why is this function here. This seems a bit overengineered to me. I would remove it.